src/mail/view/MailViewModel.ts

```diff
@@ -0,0 +1,351 @@
+			} finally {
+				this._targetMailId = null
+			}
+		} else {
+			this._targetMailId = null
+		}
+	}
+
+	async loadAndSelectMail(listId: string, mailId: string) {
+		const foundMail = await this.listModel?.loadAndSelect(
+			mailId,
+			() =>
+				// if we changed the list, stop
+				this.getListId() !== listId ||
+				// if listModel is gone for some reason, stop
+				!this.listModel ||
+				// if the target mail has changed, stop
+				this.targetMailId !== mailId ||
+				// if we loaded past the target item we won't find it, stop
+				(this.listModel.state.items.length > 0 && firstBiggerThanSecond(mailId, getElementId(lastThrow(this.listModel.state.items)))),
+		)
+		console.log("found mail", foundMail)
+	}
+
+	private async getListIdForUserInbox(): Promise<Id> {
+		const mailboxDetail = await this.mailModel.getUserMailboxDetails()
+		return assertSystemFolderOfType(mailboxDetail.folders, MailFolderType.INBOX).mails
+	}
+
+	readonly init = lazyMemoized(() => {
+		this.eventController.addEntityListener((updates) => this.entityEventsReceived(updates))
+	})
+
+	get listModel(): ListModel<Mail> | null {
+		return this._listId ? this._listModel(this._listId) : null
+	}
+
+	getMailListToSelectedMail(): ReadonlyMap<Id, Id> {
+		return this.mailListToSelectedMail
+	}
+
+	getListId(): Id | null {
+		return this._listId
+	}
+
+	private async setListId(id: Id) {
+		// Cancel old load all
+		this.listModel?.cancelLoadAll()
+		this._filterType = null
+
+		this._listId = id
+		// FIXME check if we need to unsubscribe
+		this.listModel!.stateStream.map((state) => this.onListStateChange(state))
+		await this.listModel!.loadInitial()
+	}
+
+	getConversationViewModel(): ConversationViewModel | null {
+		return this.conversationViewModel
+	}
+
+	private _listModel = memoized((listId: Id) => {
+		return new ListModel<Mail>({
+			fetch: async (startId, count) => {
+				const { complete, items } = await this.loadMailRange(listId, startId, count)
+				if (complete) {
+					this.fixCounterIfNeeded(listId, [])
+				}
+				return { complete, items }
+			},
+			loadSingle: (elementId: Id): Promise<Mail | null> => this.entityClient.load(MailTypeRef, [listId, elementId]),
+			sortCompare: sortCompareByReverseId,
+		})
+	})
+
+	private fixCounterIfNeeded: (listId: Id, itemsWhenCalled: ReadonlyArray<Mail>) => void = debounce(
+		2000,
+		async (listId: Id, itemsWhenCalled: ReadonlyArray<Mail>) => {
+			if (this._filterType != null && this.filterType !== MailFilterType.Unread) {
+				return
+			}
+
+			// If folders are changed, list won't have the data we need.
+			// Do not rely on counters if we are not connected
+			if (this.getListId() !== listId || this.connectivityModel.wsConnection()() !== WsConnectionState.connected) {
+				return
+			}
+
+			// If list was modified in the meantime, we cannot be sure that we will fix counters correctly (e.g. because of the inbox rules)
+			if (this.listModel?.state.items !== itemsWhenCalled) {
```

Comment:
would having a "commit" stream of non-nullable unknowns purely for identity comparisons like this be useful, maybe generally for view models? I don't think we have many places that need to know that the state of the model did not change when finishing some async work though

Reply:
for this case I think it already works pretty well, usually there are not so many debounced things